### PR TITLE
[Self-service error codes] Enabled by default [DPP-691]

### PR DESCRIPTION
**Pull request by tudor-da**
This PR enables self-service error codes by default in the Ledger API.
The existing `enable-self-service-error-codes` flag is replaced by `use-pre-1.18-error-codes`.

CHANGELOG_BEGIN
[Ledger API Specification] The Ledger API returns enriched error codes (see https://docs.daml.com/error-codes/self-service/index.html)
For backwards-compatibility, a new API flag `--use-pre-1.18-error-codes` is introduced for preserving the legacy behavior for clients that want to migrate incrementally to the changed gRPC status code responses and error details format.    
CHANGELOG_END

### Pull Request Checklist

- [ ] Read and understand the [contribution guidelines](https://github.com/digital-asset/daml/blob/main/CONTRIBUTING.md)
- [ ] Include appropriate tests
- [ ] Set a descriptive title and thorough description
- [ ] Add a reference to the [issue this PR will solve](https://github.com/digital-asset/daml/issues), if appropriate
- [ ] Include changelog additions in one or more commit message bodies between the `CHANGELOG_BEGIN` and `CHANGELOG_END` tags
- [ ] Normal production system change, include purpose of change in description
- [ ] If you mean to change the status of a component, please make sure you keep [the Component Status page](https://github.com/digital-asset/daml/blob/main/docs/source/support/component-statuses.rst) up to date.

NOTE: CI is not automatically run on non-members pull-requests for security
reasons. The reviewer will have to comment with `/AzurePipelines run` to
trigger the build.

### CODE REVIEW

**Review by mziolekda — APPROVED** [review]

File: ledger-service/http-json/src/itlib/scala/http/AbstractHttpServiceIntegrationTest.scala [hunk 1]
```diff
@@ -1149,8 +1149,8 @@ abstract class AbstractHttpServiceIntegrationTest
       val exerciseJson: JsValue = encodeExercise(encoder)(iouExerciseTransferCommand(contractId))
       postJsonRequest(uri.withPath(Uri.Path("/v1/exercise")), exerciseJson)
         .flatMap { case (status, output) =>
-          status shouldBe StatusCodes.Conflict
-          assertStatus(output, StatusCodes.Conflict)
+          status shouldBe StatusCodes.NotFound
```
Comment (tudor-da) [inline ledger-service/http-json/src/itlib/scala/http/AbstractHttpServiceIntegrationTest.scala:1152]: CC @stefanobaghino-da

**Review by pbatko-da — APPROVED** [review]

File: ledger/participant-state/kvutils/app/src/main/scala/com/daml/ledger/participant/state/kvutils/app/Config.scala [hunk 2]
```diff
@@ -584,11 +584,12 @@ object Config {
           else success
         )
 
-        opt[Unit]("use-self-service-error-codes")
+        opt[Unit]("use-legacy-grpc-error-codes")
```
Comment (miklos-da) [inline ledger/participant-state/kvutils/app/src/main/scala/com/daml/ledger/participant/state/kvutils/app/Config.scala:587]: Can we avoid using the word legacy, please? Why not simply have an option `disable-self-service-error-codes` or `enable-pre-1.18-error-codes`? Any of the latter would be more self-explanatory.

File: ledger/participant-state/kvutils/app/src/main/scala/com/daml/ledger/participant/state/kvutils/app/Config.scala [hunk 3]
```diff
@@ -52,7 +52,7 @@ final case class Config[Extra](
     enableMutableContractStateCache: Boolean,
     enableInMemoryFanOutForLedgerApi: Boolean,
     extra: Extra,
-    enableSelfServiceErrorCodes: Boolean,
+    useLegacyErrorCodes: Boolean,
```
Comment (miklos-da) [inline ledger/participant-state/kvutils/app/src/main/scala/com/daml/ledger/participant/state/kvutils/app/Config.scala:55]: Why rename this config parameter? We can keep it as previously and simply have the default changed + the CLI argument. You'll have a much smaller changeset.

File: ledger/sandbox-common/src/main/scala/platform/sandbox/cli/CommonCliBase.scala [hunk 4]
```diff
@@ -343,11 +343,12 @@ class CommonCliBase(name: LedgerName) {
           "Maximum command deduplication duration."
         )
 
-      opt[Unit]("use-self-service-error-codes")
+      opt[Unit]("use-legacy-grpc-error-codes")
```
Comment (miklos-da) [inline ledger/sandbox-common/src/main/scala/platform/sandbox/cli/CommonCliBase.scala:346]:
```suggestion
      opt[Unit]("disable-self-service-error-codes")
```
or
```suggestion
      opt[Unit]("use-pre-1.18-error-codes")
```

**Review by miklos-da — CHANGES_REQUESTED** [review]

File: ledger/sandbox-common/src/main/scala/platform/sandbox/cli/CommonCliBase.scala (hunk 4 above)
Comment (tudor-da) [reply ledger/sandbox-common/src/main/scala/platform/sandbox/cli/CommonCliBase.scala:346]: Hmm, I am not sold on any of those:
* `disable-self-service-error-codes` mentions self service which we deemed to keep an internal term. For the users it's just error codes
* `use-pre-1.18-error-codes` has a dot in it 🙃 and a version. It seems unusual

How about `grpc-status-codes-compat-mode`?

File: ledger/participant-state/kvutils/app/src/main/scala/com/daml/ledger/participant/state/kvutils/app/Config.scala [hunk 5]
```diff
@@ -584,11 +584,12 @@ object Config {
...
         )
 
-        opt[Unit]("use-self-service-error-codes")
+        opt[Unit]("grpc-status-codes-compatibility-mode")
           .optional()
-          .hidden()
-          .text("Enable self-service error codes.")
-          .action((_, config) => config.copy(enableSelfServiceErrorCodes = true))
+          .text(
+            "Enables legacy gRPC error code conformance mode. This option is deprecated and will be removed in future release versions."
```
Comment (tudor-da) [inline ledger/participant-state/kvutils/app/src/main/scala/com/daml/ledger/participant/state/kvutils/app/Config.scala:590]:
```suggestion
            "Enables legacy gRPC error code compatibility mode. This option is deprecated and will be removed in future release versions."
```

File: ledger/sandbox-common/src/main/scala/platform/sandbox/cli/CommonCliBase.scala [hunk 6]
```diff
@@ -343,11 +343,12 @@ class CommonCliBase(name: LedgerName) {
...
         )
 
-      opt[Unit]("use-self-service-error-codes")
+      opt[Unit]("grpc-status-codes-compatibility-mode")
         .optional()
-        .hidden()
-        .text("Enable self-service error codes.")
-        .action((_, config) => config.copy(enableSelfServiceErrorCodes = true))
+        .text(
+          "Enables legacy gRPC error code conformance mode. This option is deprecated and will be removed in future release versions."
```
Comment (tudor-da) [inline ledger/sandbox-common/src/main/scala/platform/sandbox/cli/CommonCliBase.scala:349]:
```suggestion
          "Enables legacy gRPC error code compatibility mode. This option is deprecated and will be removed in future release versions."
```

**Comment by tudor-da** [comment]
Changed to draft until the compatibility test resolution is implemented.

File: ledger/ledger-api-test-tool/src/main/scala/com/daml/ledger/api/testtool/suites/DeprecatedSandboxClassicMemoryContractKeysIT.scala [hunk 7]
```diff
@@ -0,0 +1,241 @@
...
+import daml.ledger.test.model.Test
+
+import io.grpc.Status
+
+import java.util.regex.Pattern
+
+// This test suite contains tests with assertions adapted specifically Sandbox classic with in-memory ledger backend (deprecated).
+// It asserts a more generic `INCONSISTENT` error code instead of `INCONSISTENT_CONTRACT_KEY` or `DUPLICATE_CONTRACT_KEY`
+// TODO sandbox-classic removal: Remove this tests
+final class DeprecatedSandboxClassicMemoryContractKeysIT extends LedgerTestSuite {
```
Comment (tudor-da) [inline ledger/ledger-api-test-tool/src/main/scala/com/daml/ledger/api/testtool/suites/DeprecatedSandboxClassicMemoryContractKeysIT.scala:23]: Dear reviewers, please note the explanation for this duplicated code

**Comment by tudor-da** [comment]
CC @miklos-da @mziolekda Addressed review comments and resolved the breaking compatibility test issue (see `testing.bzl`)

File: compatibility/bazel_tools/testing.bzl [hunk 8]
```diff
@@ -842,6 +843,15 @@ def sdk_platform_test(sdk_version, platform_version):
...
         extra_sandbox_next_exclusions += ["--exclude=CommandDeduplicationIT", "--additional=KVCommandDeduplicationIT"]
 
+    # Error codes are enabled by default after 1.18.0-snapshot.20211117.8399.0.a05a40ae.
+    # Before this SDK version, ledger-api-test-tool cannot correctly assert the self-service error codes.
+    # For this reason, all platforms newer than 1.18.0-snapshot.20211117.8399.0.a05a40ae will run against
+    # old ledger-api-test-tools in compatibility mode (i.e. `--grpc-status-codes-compatibility-mode`)
+    error_codes_version_enabled_by_default = "1.18.0-snapshot.20211117.8399.0.a05a40ae"
+    if versions.is_at_most(error_codes_version_enabled_by_default, sdk_version):
+        extra_sandbox_next_args += ["--grpc-status-codes-compatibility-mode"]
+        extra_sandbox_classic_args += ["--grpc-status-codes-compatibility-mode"]
```
Comment (tudor-da) [inline compatibility/bazel_tools/testing.bzl:853]: @moritzkiefer-da @garyverhaegen-da @mziolekda

File: ledger/participant-state/kvutils/app/src/main/scala/com/daml/ledger/participant/state/kvutils/app/Config.scala (hunk 3 above)
Comment (tudor-da) [reply ledger/participant-state/kvutils/app/src/main/scala/com/daml/ledger/participant/state/kvutils/app/Config.scala]: Changed it to have matching name between the config class and CLI param. Followed your advice though for less impact 👍

File: compatibility/bazel_tools/testing.bzl (hunk 8 above)
Comment (tudor-da) [reply compatibility/bazel_tools/testing.bzl:853]: Tested the compatibility tests locally with a sparser sub-set of versions.

Comment (cocreature) [reply compatibility/bazel_tools/testing.bzl:853]: queued a full run at https://dev.azure.com/digitalasset/daml/_build/results?buildId=95477&view=results. Change looks very reasonable.

Comment (tudor-da) [reply compatibility/bazel_tools/testing.bzl:853]: Failures in some 1.18 snapshots
```
### FAILURES

KVCommandDeduplicationIT
- [KVCommandDeduplicationIT:KVCommandDeduplicationSimpleDeduplicationMixedClients] Deduplicate commands within the deduplication time window using the command client and the command submission client ... Unexpected failure (StatusRuntimeException)
  ALREADY_EXISTS: Duplicate commands
- [KVCommandDeduplicationIT:KVCommandDeduplicationSimpleDeduplicationBasic] Deduplicate commands within the deduplication duration window ... Assertion failed at line 427
  assertion failed: Expecting completion with status code OK but completion has status Some(Status(6,Duplicate commands,Vector(Any(type.googleapis.com/google.rpc.ErrorInfo,<ByteString@d03020 size=26 contents="\032\030\n\017definite_answer\022\005false">,UnknownFieldSet(Map()))),UnknownFieldSet(Map())))
- [KVCommandDeduplicationIT:KVCommandDeduplicationSimpleDeduplicationCommandClient] Deduplicate commands within the deduplication time window using the command client ... Unexpected failure (StatusRuntimeException)
  ALREADY_EXISTS: Duplicate commands
```

These are failing on main and are not related to the change in the error codes. I would not block on it

**Comment by tudor-da** [comment]
Reworded commit with the changelog to reflect the new flag name ⬆️

File: ledger/ledger-api-test-tool/src/main/scala/com/daml/ledger/api/testtool/suites/DeprecatedSandboxClassicMemoryContractKeysIT.scala [hunk 9]
```diff
@@ -0,0 +1,243 @@
...
+
+import io.grpc.Status
+
+import java.util.regex.Pattern
+
+// This test suite contains some tests duplicated from ContractKeysIT,
+// but with assertions adapted specifically for Sandbox classic with in-memory ledger backend (deprecated).
+// The changed assertions assert a more generic `INCONSISTENT` error code
+// instead of `INCONSISTENT_CONTRACT_KEY` or `DUPLICATE_CONTRACT_KEY`.
+// TODO sandbox-classic removal: Remove this tests
```
Comment (tudor-da) [inline ledger/ledger-api-test-tool/src/main/scala/com/daml/ledger/api/testtool/suites/DeprecatedSandboxClassicMemoryContractKeysIT.scala:24]: Dear reviewers, please note the comments for explaining this redundant file.

File: ledger/ledger-api-test-tool/src/main/scala/com/daml/ledger/api/testtool/suites/DeprecatedSandboxClassicMemoryExceptionsIT.scala [hunk 10]
```diff
@@ -0,0 +1,65 @@
...
+import com.daml.ledger.api.testtool.infrastructure.LedgerTestSuite
+import com.daml.ledger.test.semantic.Exceptions._
+
+import io.grpc.Status
+
+// This test suite contains some tests duplicated from ExceptionsIT,
+// but with assertions adapted specifically for Sandbox classic with in-memory ledger backend (deprecated).
+// The changed assertions assert a more generic `INCONSISTENT` error code
+// instead of `INCONSISTENT_CONTRACT_KEY` or `DUPLICATE_CONTRACT_KEY`.
+// TODO sandbox-classic removal: Remove this tests
```
Comment (tudor-da) [inline ledger/ledger-api-test-tool/src/main/scala/com/daml/ledger/api/testtool/suites/DeprecatedSandboxClassicMemoryExceptionsIT.scala:19]: Dear reviewers, please note the comments for explaining this redundant file.

File: ledger/ledger-api-test-tool/src/main/scala/com/daml/ledger/api/testtool/suites/DeprecatedSandboxClassicMemoryContractKeysIT.scala (hunk 9 above)
Comment (tudor-da) [inline ledger/ledger-api-test-tool/src/main/scala/com/daml/ledger/api/testtool/suites/DeprecatedSandboxClassicMemoryContractKeysIT.scala:24]: Dear reviewers, please note the comments for explaining this redundant file.

File: ledger/sandbox-common/src/main/scala/platform/sandbox/cli/CommonCliBase.scala (hunk 4 above)
Comment (mziolekda) [reply ledger/sandbox-common/src/main/scala/platform/sandbox/cli/CommonCliBase.scala]: `use-pre-1.18-error-codes` happens to align well with our documentation. I like it.

Comment (miklos-da) [reply ledger/sandbox-common/src/main/scala/platform/sandbox/cli/CommonCliBase.scala]: Makes sense, however, with 'compatibility mode' the question is 'compatible with what?' The SDK version would make that fairly explicit.
Assuming we'll deprecate this flag for Daml SDK 2.0, I don't have a strong opinion on what we choose, any of the above should work.

File: compatibility/bazel_tools/testing.bzl (hunk 8 above)
Comment (miklos-da) [reply compatibility/bazel_tools/testing.bzl:853]: @nicu-da may have some insight into the above failures.

File: ledger/ledger-api-test-tool/src/main/scala/com/daml/ledger/api/testtool/suites/CommandServiceIT.scala [hunk 11]
```diff
@@ -329,6 +329,9 @@ final class CommandServiceIT extends LedgerTestSuite {
     }
   })
 
+  // TODO fix this test: This test is not asserting that an interpretation error is returning a stack trace.
```
Comment (miklos-da) [inline ledger/ledger-api-test-tool/src/main/scala/com/daml/ledger/api/testtool/suites/CommandServiceIT.scala:332]: Can we rather create a ticket for this, please? We'll have a better chance of fixing it.

File: ledger/ledger-api-test-tool/src/main/scala/com/daml/ledger/api/testtool/suites/DeprecatedSandboxClassicMemoryContractKeysIT.scala (hunk 9 above)
Comment (miklos-da) [inline ledger/ledger-api-test-tool/src/main/scala/com/daml/ledger/api/testtool/suites/DeprecatedSandboxClassicMemoryContractKeysIT.scala:24]:
```suggestion
// TODO sandbox-classic removal: Remove these tests
```

File: ledger/participant-state/kvutils/app/src/test/scala/com/daml/ledger/participant/state/kvutils/app/ConfigSpec.scala [hunk 12]
```diff
@@ -59,27 +59,27 @@ final class ConfigSpec
...
+        "--grpc-status-codes-compatibility-mode",
       )
     )
 
-    actual.value.enableSelfServiceErrorCodes shouldBe true
+    actual.value.enableSelfServiceErrorCodes shouldBe false
   }
 
-  it should "disable error codes v2 flag by default" in {
+  it should "enable self-service error codes by default" in {
```
Comment (miklos-da) [inline ledger/participant-state/kvutils/app/src/test/scala/com/daml/ledger/participant/state/kvutils/app/ConfigSpec.scala:74]: :+1:

File: ledger/participant-state/kvutils/app/src/main/scala/com/daml/ledger/participant/state/kvutils/app/Config.scala [hunk 13]
```diff
@@ -584,11 +584,12 @@ object Config {
...
         )
 
-        opt[Unit]("use-self-service-error-codes")
+        opt[Unit]("grpc-status-codes-compatibility-mode")
           .optional()
-          .hidden()
-          .text("Enable self-service error codes.")
-          .action((_, config) => config.copy(enableSelfServiceErrorCodes = true))
+          .text(
+            "Enables gRPC error code compatibility mode to the pre-1.18 behaviour. This option is deprecated and will be removed in future release versions."
```
Comment (miklos-da) [inline ledger/participant-state/kvutils/app/src/main/scala/com/daml/ledger/participant/state/kvutils/app/Config.scala:590]: Nit:
```suggestion
            "Enables gRPC error code compatibility mode to the pre-1.18 behaviour. This option is deprecated and will be removed in a future release."
```

File: ledger/sandbox-common/src/main/scala/platform/sandbox/cli/CommonCliBase.scala (hunk 4 above)
Comment (tudor-da) [reply ledger/sandbox-common/src/main/scala/platform/sandbox/cli/CommonCliBase.scala]: Fine then. 2ff8c8c54136c655e75bc84245fa3ea30998f5e9

**Review by miklos-da — APPROVED** [review]

File: compatibility/bazel_tools/testing.bzl (hunk 8 above)
Comment (mziolekda) [reply compatibility/bazel_tools/testing.bzl:853]: Does this need to be adopted to the the new rendition of the flags name?
```extra_sandbox_classic_args += ["--grpc-status-codes-compatibility-mode"]```

Comment (tudor-da) [reply compatibility/bazel_tools/testing.bzl]: Adapted 903a904ba787196a87042a6342574c926f4b46c1

File: ledger/ledger-api-test-tool/src/main/scala/com/daml/ledger/api/testtool/suites/CommandServiceIT.scala (hunk 11 above)
Comment (tudor-da) [reply ledger/ledger-api-test-tool/src/main/scala/com/daml/ledger/api/testtool/suites/CommandServiceIT.scala:332]: Created ticket for it

**Pull request merged**